presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -819,6 +797,7 @@ protected Type visitSubqueryExpression(SubqueryExpression node, AnalysisContext
             Type type = Iterables.getOnlyElement(descriptor.getVisibleFields()).getType();
 
             expressionTypes.put(node, type);
+            scalarSubqueries.add(node);
```

Comment:
As we discussed we should change visitSubqueryExpression so it is aware of the position of the subquery being analysed in expression tree.

Then different checks would be done for 
1) subqueries which are right hand side of `IN` operator. We require single column queries for it.
2) subqueries which are used as scalar value (e.g. `x > (SELECT ...). Here we require scalar subquery.

And then different subqueries groups would be stored on different fields in `ExpressionAnalysis` object.